core/combinedvideosource.go

```diff
@@ -0,0 +1,283 @@
+package core
+
+import (
+	"sync"
+
+	"github.com/ericxtang/m3u8"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/go-livepeer/net"
+	ffmpeg "github.com/livepeer/lpms/ffmpeg"
+	"github.com/livepeer/lpms/stream"
+)
+
+var liveListLev uint = 10
+
+// CombinedVideoSource holds data in different OSes
+type CombinedVideoSource struct {
+	sources      map[ManifestID]*masterSource
+	sourcesPLock sync.Mutex
+}
+
+type masterSource struct {
+	manifestID         ManifestID
+	masterLivePList    *m3u8.MasterPlaylist
+	masterFullPList    *m3u8.MasterPlaylist
+	streams            []*videoStream
+	masterManifestName string
+}
+
+type videoStream struct {
+	streamID          StreamID
+	osSession         drivers.OSSession
```

Comment:
Why do we have osSession on the `videoStream` level instead of on the `masterSource` level?  Do we need the flexibility to allow video streams in the same MasterPlaylist to use different storage methods?

Reply:
It was whole point of OS proposal - to allow different use of different OSes in different combinations. With this design we can write streams into S3, into local storage (plus saving it to disk), and into IPFS, simultaneously.